### PR TITLE
Mancer API

### PR DESCRIPTION
**Pull request by 50h100a**
[I actually went and did it.](https://mancer.tech/)

The site's awful, but the service should work.

Using `X-API-KEY` header value for authentication.
Not supporting `token-count`, I think it's a bad approach with remote APIs anyway, unless we add batching.
Not supporting streaming generation (yet), it's just not part of the MVP.
Made no changes to `migrateSecrets`, as I didn't think it was relevant.

New WebUI connection screen:
firefox_s5mjWkQn2F

Now with API key:
firefox_FHJHTyXaYQ

We could make the API-key wording generic, but I don't know of any other services providing a generic webui-like interface, so...

### CODE REVIEW

File: public/script.js [hunk 1]
```diff
@@ -7611,16 +7624,28 @@ $(document).ready(function () {
         }
     });
 
+    $("#use-mancer-api-checkbox").on("change", function (e) {
+        const enabled = $("#use-mancer-api-checkbox").prop("checked");
+        $("#mancer-api-ui").css("display", enabled?"block":"none");
```
Comment (Cohee1207) [inline public/script.js]: Could be replaced with `$('#mancer-api-ui').toggle(enabled)`

File: public/script.js [hunk 2]
```diff
@@ -7611,16 +7624,28 @@ $(document).ready(function () {
...
-                callPopup('Please enter a valid URL.', 'text');
+            let value = $("#textgenerationwebui_api_url_text").val().trim();
+            if (!value || !value.endsWith('/api')) {
+                callPopup('Please enter a valid URL.<br/>WebUI URLs should end with <tt>/api</tt>', 'text');
                 return;
             }
 
+            const mancer_key = $("#api_key_mancer").val().trim();
+            if (mancer_key.length) {
+                writeSecret(SECRET_KEYS.MANCER, mancer_key);
```
Comment (Cohee1207) [inline public/script.js]: await writeSecret call before proceeding (it's async)

File: public/script.js [hunk 3]
```diff
@@ -7611,16 +7624,28 @@ $(document).ready(function () {
...
     $("#api_button_textgenerationwebui").click(function (e) {
         e.stopPropagation();
         if ($("#textgenerationwebui_api_url_text").val() != "") {
-            let value = formatKoboldUrl($("#textgenerationwebui_api_url_text").val().trim());
-
-            if (!value) {
-                callPopup('Please enter a valid URL.', 'text');
+            let value = $("#textgenerationwebui_api_url_text").val().trim();
+            if (!value || !value.endsWith('/api')) {
+                callPopup('Please enter a valid URL.<br/>WebUI URLs should end with <tt>/api</tt>', 'text');
```
Comment (Cohee1207) [inline public/script.js:7706]: I think some form of URL checking should stay (but without forcing /api path like formatKoboldUrl did)
`pancakes/api` would be considered valid by this code

File: public/script.js [hunk 4]
```diff
@@ -2933,6 +2935,14 @@ async function Generate(type, { automatic_trigger, force_name2, resolve, reject,
                     activateSendButtons();
                     //console.log('runGenerate calling showSwipeBtns');
                     showSwipeButtons();
+
+                    if (main_api == 'textgenerationwebui' && api_use_mancer_webui) {
+                        const errorText = `<h3>Inferencer endpoint is unhappy!</h3>
```
Comment (Cohee1207) [inline public/script.js:2991]: I love the silly text but it is not informative enough. Maybe it is worth trying to pass a status code at least

File: public/index.html [hunk 5]
```diff
@@ -1501,30 +1501,55 @@ <h4><span data-i18n="Novel AI Model">Novel AI Model</span>
...
+                                    Mancer AI
+                                </a>, 
+                                <label class="checkbox_label" for="use-mancer-api-checkbox">
+                                    <span data-i18n="Use API key (Only required for Mancer)">
+                                        Click this box (and add your API key!):
+                                    </span>
+                                    <input id="use-mancer-api-checkbox" type="checkbox" />
+                                </label>
+                            </div>
+                            <div id="mancer-api-ui" style="display:none;">
```
Comment (Cohee1207) [inline public/index.html:1563]: I think displayNone class should do the trick instead of raw style

File: public/index.html [hunk 6]
```diff
@@ -1501,30 +1501,55 @@ <h4><span data-i18n="Novel AI Model">Novel AI Model</span>
...
-                        </div>
-                        <div>
-                            <div class="flex-container flexFlowColumn">
-                                <div class="flex1">
-                                    <h4 data-i18n="Blocking API url">Blocking API url</h4>
-                                    <small>Example: http://127.0.0.1:5000/</small>
-                                    <input id="textgenerationwebui_api_url_text" name="textgenerationwebui_api_url" class="text_pole wide100p" maxlength="500" value="" autocomplete="off">
+                        <form action="javascript:void(null);" method="post" enctype="multipart/form-data">
+                            If you are using: 
+                            <div class="flex-container" style="margin-left:20px;">
```
Comment (Cohee1207) [inline public/index.html]: Try defining a style rule instead of hardcoded margin

File: public/script.js (hunk 1 above)
Comment (50h100a) [reply public/script.js]: sorcery!

File: public/script.js (hunk 4 above)
Comment (50h100a) [reply public/script.js:2991]: Is the problem with the title specifically, or just to make the description more specific?

File: public/index.html (hunk 5 above)
Comment (50h100a) [reply public/index.html:1563]: tried, but the `!important` seems to break the toggling.

File: public/script.js (hunk 4 above)
Comment (Cohee1207) [reply public/script.js:2991]: More specific, that may help you debug problems easier

**Comment by Cohee1207** [comment]
Don't forget to add your API to https://docs.sillytavern.app/usage/api-connections/
Small description + a link to your site is ok

**Comment by 50h100a** [comment]
I'll add something to the proper docs once I've made sure things more-or-less work with a smaller crowd.

**Pull request merged**